jme3-core/src/main/resources/Common/ShaderLib/PBRLighting.glsllib

```diff
@@ -0,0 +1,184 @@
+#import "Common/ShaderLib/Lighting.glsllib"
+#import "Common/ShaderLib/PBR.glsllib"
+
+//declare PBR Lighting vars
+uniform vec4 g_LightData[NB_LIGHTS];
+uniform vec3 g_CameraPosition;
+uniform vec4 g_AmbientLightColor;
+
+#if NB_PROBES >= 1
+    uniform samplerCube g_PrefEnvMap;
+    uniform vec3 g_ShCoeffs[9];
+    uniform mat4 g_LightProbeData;
+#endif
+#if NB_PROBES >= 2
+    uniform samplerCube g_PrefEnvMap2;
+    uniform vec3 g_ShCoeffs2[9];
+    uniform mat4 g_LightProbeData2;
+#endif
+#if NB_PROBES == 3
+    uniform samplerCube g_PrefEnvMap3;
+    uniform vec3 g_ShCoeffs3[9];
+    uniform mat4 g_LightProbeData3;
+#endif
+
+// Specular-AA
+#ifdef SPECULAR_AA_SCREEN_SPACE_VARIANCE
+    uniform float m_SpecularAASigma;
+#endif
+#ifdef SPECULAR_AA_THRESHOLD
+    uniform float m_SpecularAAKappa;
+#endif
+
+vec3 calculatePBRLighting(in vec4 albedo, in float Metallic, in float Roughness, in vec4 specularColor, in float glossiness, in vec3 lightMapColor, in vec3 ao, in float indoorSunLightExposure, in vec3 normal, in vec3 norm, in vec3 viewDir){

```

Comment:
This code can be refactored into generic functions in PBR.glsllib.
To achieve this, all the code that relies on preprocessor directives (eg. #ifdef LIGHTMAP , for( int i = 0;i < NB_LIGHTS; i+=3){ ) should be split into different functions that are then called by the material shader depending on its logic (eg. a single pass lighting would use the for-loop NB_LIGHTS, but a multipass shader would not, so having this logic baked calculatePBRLighting  limits its flexibility).

Reply:
Making a single-pass version of the PBRLighting shader may be out of the scope of my capabilities currently, since (unless I'm mistaken) the current PBRLighting implementation only handles multi-pass and I'm mostly learning from working with that code, so I was mostly writing the PBRLighting.glsllib to be just be a multi-pass implementation

So maybe I should rename PBRLighting.glsllib to instead be MultiPassPbrLighting.glsllib ?

(and then I can still try split the non-multi pass lighting code into more modular functions that could also be used by a future single pass implementation that would be called SinglePassPBRLighting.glsllib)

Reply:
The current PBRLighting is single pass. The point i was making is that in PBRLighting.glsllib there are parts that are generic PBR stuff, parts that are for light calculation in world space and parts that are just specific to the PBRLighting implementation.

Even if you split the code as you did, it solves only some of the issue, eg. if you want to implement a shader that uses UBOs (now that they are core) you would need to rewrite the PBRLighting.glsllib entirely, that's why a further abstraction would be preferable.

Eg. see this code that uses structs and UBOs for dynamic lighting in tiled deferred shading

```glsl
            vec3 dlightSum=vec3(0);
            for(int i=0;i<MAX_LIGHTS_PER_GEOM;i++){
                // this part is specific to the implementation and collects the light data passed as UBO that you normally have in g_LightData in jme
                if(i>=bo_LightsData.nLights||i>=m_NLightsPerGeo)break;
                int lightIndex=m_LightsIndices[i];
                Light light=bo_LightsData.lights[lightIndex];
              
                // This is the generic world space light computation that you have in PBRLighting.glsllib
                ComputedJmeLight clight=Light_computeFor(surface.position, surface.normal, surface.viewDir, light);

                // This is the PBR lighting 
                vec3 lightResult = PBR_computeDirectLight(surface,clight);    
               
                
                dlightSum += lightResult;
            }
            outColor.rgb+=dlightSum;
```

This would go into the fragment shader and as you can see it can reuse most of the code and abstracts away all the complex logic.

Reply:
I could have swore that I was told before that (even though the technique is called SinglePassAndImageBased) that PBRLighting is actually multi-pass and has to do another pass to re-draw the scene once per light (and I thought that was the reason why my scenes lag with many point lights, and why a deferred rendering pipeline is needed to instead do the passes more efficiently).

Is this assumption incorrect?

I could also try my best to refactor the contents of PBRLighting.glsllib to be reusable to support more lighting techniques, but since there isn't really an official deferred rendering pipeline in jme-core (and I'm far from an expert in the subject) I really don't think I will be able to do a good job, since I've only barely messed with the lighting code in the current pipeline and can't say I fully understand it currently.

I had planned to wait til jhonkkk was done with his deferred rendering pipeline before doing this this PR (briefly discussed in the issue that spurred this PR: https://github.com/jMonkeyEngine/jmonkeyengine/issues/2122) and I was going to use his pipeline as a reference to compare the current pipeline to so I could pull out and refactor duplicate code, but now that he disappeared I figured I'd atleast go ahead and modularize it as best as I can for JME's one current rendering pipeline.

So maybe I should focus on modularizing the ParamReads glsllib and leave the PBRLighting.glsllib alone for now, atleast until we finally get a second rendering pipeline in the engine. Then once I have gained experience with a different type of pipeline and its lighting code I could probably come back and properly modularize the lighting side of things. But for right now that feels beyond my capabilities.

Reply:
> I could have swore that I was told before that (even though the technique is called SinglePassAndImageBased) that PBRLighting is actually multi-pass and has to do another pass to re-draw the scene once per light (and I thought that was the reason why my scenes lag with many point lights, and why a deferred rendering pipeline is needed to instead do the passes more efficiently).
>Is this assumption incorrect?

Yes, it is incorrect, PBRLighting is single pass, but there is a catch: you can't efficiently render unlimited lights with a fragment shaders, so jme does batch them in an array up to `renderManager.getSinglePassLightBatchSize()` in size (nb. you can set any value for it but bigger values will consume more vram and bandwidth) and then renders each batch with one pass each.

Eg. 
- SinglePassLightBatchSize = 64, you have 32 visible lights in your scene - > scene is rendered in 1 pass
- SinglePassLightBatchSize = 64, you have 65 visible lights in your scene - > scene is rendered in 2 passes
- SinglePassLightBatchSize = 64, you have 80 visible lights in your scene - > scene is rendered in 2 passes (still)
- SinglePassLightBatchSize = 64, you have 129 visible lights - > scene is rendered in 3 passes

You see what i mean.

There are better way to do it, such as tiled deferred rendering, where lights are rendered in one or more passes for each tile of the screen and then shading is done singlepass after.



> So maybe I should focus on modularizing the ParamReads glsllib and leave the PBRLighting.glsllib as it is for now, atleast until we finally get a second rendering pipeline in the engine. Then once I have gained experience with a different type of pipeline and its lighting code I could probably come back and properly modularize the lighting side of things. But for right now that feels beyond my capabilities.

Mhh i'd like to hear your opinion on this: could we tackle this issue the other way around?
Eg. keep PBRLighting.frag as a single file (maybe reorganized a bit as you did) and have  a way to inject/hook custom code into it?

I am thinking for example: i want to add a code in PBRLighting to do some computation after the parameters are read, the PBRLighting.frag will have something like

```glsl
... something something...
vec3 normal=texture(......
... something something...

// THE HOOK
#ifdef _HOOK_POST_PARAM_READ_
_HOOK_POST_PARAM_READ_(param1, normal, param3...)
#endif

```

and in MyForkPBRLighting.frag i will have something like

```glsl
// create transform function
void postParamReadHook(inout param1, inout normal, inout param3....){
    normal*=2;
}
// hook into PBRLighting
#define _HOOK_POST_PARAM_READ_ postParamReadHook

#import "PBRLighting.frag"
```

in this way i can really extend PBRLighting.frag from my shader in a more "standard" way, by hooking into specific parts of the code, we can have many hooks for each section with no performance impact whatsoever and with very minimal refactoring.

But we would still need the uniforms and varying to be into a separate glsllib, that can be imported multiple times without breaking the shader, so since we don't have `pragma once;`  in glsl, it should use macros like if we were C programmers in the 70s:

```glsl
#ifndef __PBR_PARAMS__
#define __PBR_PARAMS__ 1
    uniform sampler2D m_.....
    ......
    ......
#endif

```
So PBRLighting.frag could import it internally but if need the fork can also import the same uniform and varying to be used into postParamReadHook by just importing the file at the top, first import will win.

What do you think?

Reply:
> * SinglePassLightBatchSize = 64, you have 32 visible lights in your scene - > scene is rendered in 1 pass

Hmm this still confuses me based on what I thought I'd been told in the past and especially in regards to my own findings when trying to use even a few point lights. It sounds like 9based on what you're saying) that I should be able to use PBRLighting with up to 64 lights without having framerate issues since its all in one pass under 64, but I can barely even use 2 point lights with in a moderate sized scene.

Pretty much all my scenes (even ones with just a few triplanar terrains and no other models) can't render even 1-3 PointLights without destroying the framerate because (I thought) that every additional light was causing it to render the terrain and do all the triplanar texture reads again for each light. 

Is that not correct? And if so then why am I getting such bad lag with only 1-3 point lights if it is <64 and supposedly all done in one pass? Is it something else causing the lights to lag then? 

___

I also am not quite sure if I understand the hooking concept or the benefit it would add. It seems more confusing to me than if I were to just call a method from a glsllib between the param reads and lighting calculation and (unless I'm misunderstanding) it seems like the hooking could end up in a situation where there is conflicting variables since PBRLighting.frag is imported after the custom code. I would expect the hooking example code you pasted to have more issues with conflicting variables since the PBRLighting.frag import is declared after defining the hook (so the forked shader won't know about any of the variables in PBRLighting.frag, so then PBRLighting would need all the not-defined checks and would just send up with a bunch of extra unnecessary code checking if certain uniforms were already defined or not.

So I don't quite understand how the hooking would be any different than the way I currently did it, both just seem like different way to put code between the param reads and lighting, but hooking seems like it would just require more code to get it to work by checking the #ifndef before everyuniform in PBRLighting.frag

I still don't see the issue with just defining every uniform and variable that has potential to be re-used in the .frag shader. Then as long as the user puts the import for their glsllib after the variable declarations in PBRLighting.frag (which already is required in some cases even without this PR), then they will never have to worry about double declarations.

It seems like this would be a lot of work just to protect a potentially careless shader dev from accidentally declaring a variable twice, which IMO isn't a big worry becuase in the case that a glsllib does try to declare an already existent uniform, the app will just crash and tell them which variable is already declared, then it should only take a minute to go back into their glsllib and remove the uniform that is already declared in pbrlighting.frag, or add the #nifdef to their glsllib (rather than putting those not-defined checks all in PBRLighting.frag). So it sort of feels like this would a lot of extra code in PBRLighting.frag just to add guard-rails to prevent a mistake that is easy to fix.

So unless I'm still overlooking something, it feels like the way I have it set up is just simpler and will do the same thing with less code.

Reply:
> Eg. keep PBRLighting.frag as a single file (maybe reorganized a bit as you did)

This would also defeat the purpose of this PR since I also plan to reuse the PBRLigthing.glsllib for the lighting calculations in PBRTerrain.frag and AdvancedPBRTerrain.frag in the jme-terrain library. So I will also be refactoring both pbr terrain shaders in jme-terrain to work with this PR and share a lighting .glsllib.

And that was where the idea originally came idea too because jhonkkk was adapting PBRLighting + both terrain shaders to his deferred rendering, and he noticed they all share the same lighting code and wanted to be able to make changes in one single glsllib (rather than changing all 3 .frag shaders) to develop all 3 easier and ensure they were consistent.

Reply:
> > * SinglePassLightBatchSize = 64, you have 32 visible lights in your scene - > scene is rendered in 1 pass
> 
> Hmm this still confuses me based on what I thought I'd been told in the past and especially in regards to my own findings when trying to use even a few point lights. It sounds like (based on what you're saying) that I should be able to use PBRLighting with up to 64 lights without having framerate issues since its all in one pass for up to 64 lights, but I can barely even use 2 point lights with in a moderate sized scene without the framerate dropping drastically.
> 
> Pretty much all my scenes (even ones with just a few triplanar terrains and no other models) can't render even 1-3 PointLights without destroying the framerate because (I thought) that every additional light was causing it to render the terrain and do all the triplanar texture reads again for each light.
> 
> Is that not correct? And if so then why am I getting such bad lag with only 1-3 point lights if it is <64 and supposedly all done in one pass? Is it something else causing the lights to lag then?

I said 64 as an example, singlePassLightBatchSize is actually 1 by default, you should set it with `renderManager.setSinglePassLightBatchSize(int)` to a value that make sense for your scene.

> I also am not quite sure if I understand the hooking concept or the benefit it would add. It seems more confusing to me than if I were to just call a method from a glsllib between the param reads and lighting calculation and (unless I'm misunderstanding) it seems like the hooking could end up in a situation where there is conflicting variables since PBRLighting.frag is imported after the custom code. I would expect the hooking example code you pasted to have more issues with conflicting variables since the PBRLighting.frag import is declared after defining the hook (so the forked shader won't know about any of the variables in PBRLighting.frag, so then PBRLighting would need all the not-defined checks and would just send up with a bunch of extra unnecessary code checking if certain uniforms were already defined or not.
> 
> So I don't quite understand how the hooking would be any different than the way I currently did it, both just seem like different way to put code between the param reads and lighting, but hooking seems like it would just require more code to get it to work by checking the #ifndef before everyuniform in PBRLighting.frag, and would also add more defines for each hook (which could be problematic currently since the current version of jme limits each shader to 64 defines until another PR fixing that issue is merged to stable, so I always typically avoid creating new defines as much as possible)
> 
> I still don't see the issue with just defining every uniform and variable that has potential to be re-used in the .frag shader. Then as long as the user puts the import for their glsllib after the variable declarations in PBRLighting.frag (which already is required in some cases even without this PR), then they will never have to worry about double declarations.
> 
> It seems like this would be a lot of work just to protect a potentially careless shader dev from accidentally declaring a variable twice, which IMO isn't a big worry becuase in the case that a glsllib does try to declare an already existent uniform, the app will just crash and tell them which variable is already declared, then it should only take a minute to go back into their glsllib and remove the uniform that is already declared in pbrlighting.frag, or add the #nifdef to their glsllib (rather than putting those not-defined checks all in PBRLighting.frag). So it sort of feels like this would a lot of extra code in PBRLighting.frag just to add guard-rails to prevent a mistake that is easy to fix.
> 
> So unless I'm still overlooking something, it feels like the way I have it set up is just simpler and will do the same thing with less code. I also don't know if there would be any benefit to putting more hooks in more places (aside from between the lighting and param reads), since there should almost never be any reason to change the way that the base PBR textures and params are read, otherwise you'd be breaking PBRLighting and it really wouldn't be a fork of PBRLighting anymore.
> 
> My thoughts were just to put the param reads and lighting into 2 big glsllibs since the code in those libs should almost never change for a true fork of PBRLighting, and the lighting code would only change if you switch to a different lighting method at which point you can just use a different glsllib for the lighting part. I think anything further than that needlessly overcomplicates it and will juts make existing forks require even more refactoring to get up to date with the changes to use the hooking method.



> > Eg. keep PBRLighting.frag as a single file (maybe reorganized a bit as you did)
> 
> This would also defeat the purpose of this PR since I also plan to reuse the PBRLigthing.glsllib for the lighting calculations in PBRTerrain.frag and AdvancedPBRTerrain.frag in the jme-terrain library. So I will also be refactoring both pbr terrain shaders in jme-terrain to work with this PR and share a lighting .glsllib.
> 
> And that was where the idea originally came idea too because jhonkkk was adapting PBRLighting + both terrain shaders to his deferred rendering, and he noticed they all share the same lighting code and wanted to be able to make changes in one single glsllib (rather than changing all 3 .frag shaders) to develop all 3 easier and ensure they were consistent.
> 
> There would also be nothing stopping a jme dev from making a fork where they put the code from the 2 glsllibs back into one mega .frag shader if that is necessary for their fork somehow or if it feels easier for them.

Ok, i'll think more about this use case, can you show (even in pseudocode) how PBRTerrain can be refactored with this PR? Do you need to include the two glsllib before and after the terrain code?

Reply:
PBRTerrain would have its texture reads moved to another .glsllib called something like PBRTerrainParmReader, and that would also be reusable by another lighting implementation such as the one that jhonkkk made based on PBRTerrain:

https://github.com/jMonkeyEngine/jmonkeyengine/blob/9e977c67536baa9ac1625c80b868d776a1088e1f/jme3-terrain/src/main/resources/Common/MatDefs/Terrain/GBufferPack/PBRTerrainGBufferPack.frag

So once all the refactoring was done, his PBRTerrain shader's main function would look like:

```
readPBRTerrainParams():
calculateGBufferPackLightingStuff();

```

and the original PBRTerrain shader currently in jme-terrain would use the same texture-reading .glsllib, but would use the calculateFinaLightingValue() method that is also in PBRLighting.frag. So PBRterrain's main funciton would look like:

```
readPBRTerrainParams();
vec4 finalLightingValue = calculateFinalLightingValue();

```

and PBRLighting.frag looks like:

```
readPBRTParams();
vec4 finalLightingValue = calculateFinalLightingValue();

```

and jhonkkk's gbufferpack implementation (PBRLightingGBufferPack.frag) looks like:

```
readPBRTParams();
calculateGBufferPackLightingStuff();


So PBRligthing, PBRTerrain, and AdvacnedPBRTerrain will all use the lighting code from PBRLighting.glsllib. 

And then another lighting implementation (such as the gbufferpack-based one from jhonkkk that I linked) would re-use the glsllibs for reading params, but would all share a different GBufferPackLighting.glslib file.

So if I had done this PR earlier, then this would have drastically reduced the work that someone such as jhonkkk would've had to do when he had to copy/paste all the param-reading code from PBRLighting. PBRTerrain, and AdvancedPBRTerrain ino his own GBufferpack implementations

Reply:
> I said 64 as an example, singlePassLightBatchSize is actually 1 by default, you should set it with `renderManager.setSinglePassLightBatchSize(int)` to a value that make sense for your scene.

I never knew that was something that could be increased with the current pipeline, thanks! I'll have to try that and hopefully I can start using some point lights again in my game.